File: pricing/pricing/page/quick_quotation/quick_quotation.js
frappe.pages['quick-quotation'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'قیمت‌گذاری سریع',
        single_column: true
    });
    
    // ایجاد instance از کلاس صفحه
    new QuickQuotation(page);
};

class QuickQuotation {
    constructor(page) {
        this.page = page;
        this.selected_items = [];
        this.settings = {};
        
        this.make();
        this.setup_buttons();
        this.load_settings();
    }
    
    make() {
        this.page.main.html(`
            <div class="quick-quotation-container" style="padding: 20px; direction: rtl;">
                <!-- بخش جستجو و اضافه کردن کالا -->
                <div class="row" style="margin-bottom: 20px;">
                    <div class="col-md-8">
                        <div class="frappe-control" id="item-search-wrapper"></div>
                    </div>
                    <div class="col-md-2">
                        <div class="frappe-control" id="qty-wrapper"></div>
                    </div>
                    <div class="col-md-2">
                        <button class="btn btn-primary btn-block" id="add-item-btn">
                            <i class="fa fa-plus"></i> افزودن
                        </button>
                    </div>
                </div>
                
                <!-- جدول کالاهای انتخاب شده -->
                <div class="selected-items-section" style="margin-bottom: 20px;">
                    <h4 style="margin-bottom: 15px; border-bottom: 2px solid #5e64ff; padding-bottom: 10px;">
                        <i class="fa fa-shopping-cart"></i> کالاهای انتخاب شده
                    </h4>
                    <div id="items-table-wrapper">
                        <table class="table table-bordered table-hover" id="items-table">
                            <thead style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white;">
                                <tr>
                                    <th style="width: 40px;">#</th>
                                    <th>کد کالا</th>
                                    <th>نام کالا</th>
                                    <th style="width: 80px;">تعداد</th>
                                    <th>بهای تمام شده</th>
                                    <th>قیمت فروش</th>
                                    <th>سود</th>
                                    <th>قیمت با تخفیف</th>
                                    <th>قیمت قسطی</th>
                                    <th style="width: 60px;">حذف</th>
                                </tr>
                            </thead>
                            <tbody id="items-tbody">
                                <tr id="empty-row">
                                    <td colspan="10" style="text-align: center; color: #999; padding: 30px;">
                                        <i class="fa fa-info-circle"></i> کالایی انتخاب نشده است
                                    </td>
                                </tr>
                            </tbody>
                        </table>
                    </div>
                </div>
                
                <!-- بخش خلاصه و جمع کل -->
                <div class="summary-section" style="background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%); padding: 20px; border-radius: 10px; margin-bottom: 20px;">
                    <div class="row">
                        <div class="col-md-3">
                            <div class="summary-card" style="background: white; padding: 15px; border-radius: 8px; text-align: center; box-shadow: 0 2px 10px rgba(0,0,0,0.1);">
                                <h5 style="color: #666; margin-bottom: 5px;">مجموع بهای تمام شده</h5>
                                <h3 id="total-cost" style="color: #e74c3c; margin: 0;">0</h3>
                            </div>
                        </div>
                        <div class="col-md-3">
                            <div class="summary-card" style="background: white; padding: 15px; border-radius: 8px; text-align: center; box-shadow: 0 2px 10px rgba(0,0,0,0.1);">
                                <h5 style="color: #666; margin-bottom: 5px;">مجموع قیمت فروش</h5>
                                <h3 id="total-selling" style="color: #3498db; margin: 0;">0</h3>
                            </div>
                        </div>
                        <div class="col-md-3">
                            <div class="summary-card" style="background: white; padding: 15px; border-radius: 8px; text-align: center; box-shadow: 0 2px 10px rgba(0,0,0,0.1);">
                                <h5 style="color: #666; margin-bottom: 5px;">مجموع سود</h5>
                                <h3 id="total-profit" style="color: #27ae60; margin: 0;">0</h3>
                            </div>
                        </div>
                        <div class="col-md-3">
                            <div class="summary-card" style="background: white; padding: 15px; border-radius: 8px; text-align: center; box-shadow: 0 2px 10px rgba(0,0,0,0.1);">
                                <h5 style="color: #666; margin-bottom: 5px;">مجموع قسطی</h5>
                                <h3 id="total-installment" style="color: #9b59b6; margin: 0;">0</h3>
                            </div>
                        </div>
                    </div>
                </div>
                
                <!-- جزئیات قسط -->
                <div class="installment-details" id="installment-details" style="background: #fff3cd; padding: 15px; border-radius: 8px; margin-bottom: 20px; display: none;">
                    <h5 style="color: #856404; margin-bottom: 10px;">
                        <i class="fa fa-calculator"></i> جزئیات پرداخت قسطی
                    </h5>
                    <div class="row">
                        <div class="col-md-3">
                            <strong>پیش پرداخت:</strong> <span id="detail-down-payment">0</span>
                        </div>
                        <div class="col-md-3">
                            <strong>قسط ماهانه:</strong> <span id="detail-monthly-payment">0</span>
                        </div>
                        <div class="col-md-3">
                            <strong>تعداد اقساط:</strong> <span id="detail-months">0</span> ماه
                        </div>
                        <div class="col-md-3">
                            <strong>مجموع بهره:</strong> <span id="detail-total-interest">0</span>
                        </div>
                    </div>
                </div>
                
                <!-- دکمه‌های عملیات -->
                <div class="action-buttons" style="text-align: left;">
                    <button class="btn btn-danger" id="clear-all-btn" style="margin-left: 10px;">
                        <i class="fa fa-trash"></i> پاک کردن همه
                    </button>
                    <button class="btn btn-success" id="create-quotation-btn">
                        <i class="fa fa-file-text"></i> ایجاد پیش‌فاکتور
                    </button>
                    <button class="btn btn-info" id="print-btn" style="margin-right: 10px;">
                        <i class="fa fa-print"></i> چاپ
                    </button>
                </div>
            </div>
        `);
        
        this.setup_fields();
        this.setup_events();
    }
    
    setup_fields() {
        // فیلد جستجوی کالا
        this.item_search = frappe.ui.form.make_control({
            df: {
                fieldname: 'item_code',
                fieldtype: 'Link',
                options: 'Item',
                label: 'جستجوی کالا',
                placeholder: 'کد یا نام کالا را وارد کنید...'
            },
            parent: this.page.main.find('#item-search-wrapper'),
            render_input: true
        });
        
        // فیلد تعداد
        this.qty_field = frappe.ui.form.make_control({
            df: {
                fieldname: 'qty',
                fieldtype: 'Float',
                label: 'تعداد',
                default: 1
            },
            parent: this.page.main.find('#qty-wrapper'),
            render_input: true
        });
        this.qty_field.set_value(1);
    }
    
    setup_events() {
        let me = this;
        
        // دکمه افزودن کالا
        this.page.main.find('#add-item-btn').on('click', () => {
            me.add_item();
        });
        
        // Enter در فیلد جستجو
        this.page.main.find('#item-search-wrapper input').on('keypress', function(e) {
            if (e.which == 13) {
                me.add_item();
            }
        });
        
        // دکمه پاک کردن همه
        this.page.main.find('#clear-all-btn').on('click', () => {
            me.clear_all();
        });
        
        // دکمه ایجاد پیش‌فاکتور
        this.page.main.find('#create-quotation-btn').on('click', () => {
            me.create_quotation();
        });
        
        // دکمه چاپ
        this.page.main.find('#print-btn').on('click', () => {
            me.print_quotation();
        });
    }
    
    setup_buttons() {
        // دکمه تنظیمات
        this.page.set_secondary_action('تنظیمات', () => {
            this.show_settings_dialog();
        }, 'octicon octicon-gear');
    }
    
    load_settings() {
        // بارگذاری تنظیمات از آخرین Auto Price List
        frappe.call({
            method: 'pricing.pricing.page.quick_quotation.quick_quotation.get_auto_price_lists',
            callback: (r) => {
                if (r.message && r.message.length > 0) {
                    this.settings = r.message[0];
                }
            }
        });
    }
    
    add_item() {
        let item_code = this.item_search.get_value();
        let qty = this.qty_field.get_value() || 1;
        
        if (!item_code) {
            frappe.show_alert({
                message: 'لطفاً کالا را انتخاب کنید',
                indicator: 'orange'
            });
            return;
        }
        
        // بررسی تکراری نبودن
        let existing = this.selected_items.find(i => i.item_code === item_code);
        if (existing) {
            existing.qty += qty;
            this.refresh_item_prices();
            return;
        }
        
        // دریافت قیمت کالا
        frappe.call({
            method: 'pricing.pricing.page.quick_quotation.quick_quotation.get_item_price_details',
            args: {
                item_code: item_code,
                qty: qty
            },
            callback: (r) => {
                if (r.message && !r.message.error) {
                    this.selected_items.push(r.message);
                    this.render_items_table();
                    this.update_totals();
                    
                    // پاک کردن فیلدها
                    this.item_search.set_value('');
                    this.qty_field.set_value(1);
                    this.item_search.$input.focus();
                } else {
                    frappe.show_alert({
                        message: r.message.error || 'خطا در دریافت قیمت',
                        indicator: 'red'
                    });
                }
            }
        });
    }
    
    render_items_table() {
        let tbody = this.page.main.find('#items-tbody');
        tbody.empty();
        
        if (this.selected_items.length === 0) {
            tbody.html(`
                <tr id="empty-row">
                    <td colspan="10" style="text-align: center; color: #999; padding: 30px;">
                        <i class="fa fa-info-circle"></i> کالایی انتخاب نشده است
                    </td>
                </tr>
            `);
            return;
        }
        
        let me = this;
        this.selected_items.forEach((item, index) => {
            let prices = item.prices || {};
            let installment = item.installment || {};
            
            let row = $(`
                <tr data-index="${index}">
                    <td>${index + 1}</td>
                    <td><strong>${item.item_code}</strong></td>
                    <td>${item.item_name || ''}</td>
                    <td>
                        <input type="number" class="form-control qty-input" 
                               value="${item.qty}" min="1" style="width: 70px; text-align: center;">
                    </td>
                    <td style="color: #e74c3c;">${this.format_currency(prices.total_cost)}</td>
                    <td style="color: #3498db; font-weight: bold;">${this.format_currency(prices.total_selling_price)}</td>
                    <td style="color: ${prices.total_profit >= 0 ? '#27ae60' : '#e74c3c'};">
                        ${this.format_currency(prices.total_profit)}
                        <small>(${(prices.profit_percentage || 0).toFixed(1)}%)</small>
                    </td>
                    <td style="color: #8e44ad;">${this.format_currency(prices.total_price_with_markup || prices.total_selling_price)}</td>
                    <td style="color: #9b59b6;">${this.format_currency(installment.total_amount || 0)}</td>
                    <td>
                        <button class="btn btn-danger btn-xs remove-item-btn">
                            <i class="fa fa-times"></i>
                        </button>
                    </td>
                </tr>
            `);
            
            // رویداد تغییر تعداد
            row.find('.qty-input').on('change', function() {
                item.qty = parseFloat($(this).val()) || 1;
                me.refresh_item_prices();
            });
            
            // رویداد حذف
            row.find('.remove-item-btn').on('click', function() {
                me.selected_items.splice(index, 1);
                me.render_items_table();
                me.update_totals();
            });
            
            tbody.append(row);
        });
    }
    
    refresh_item_prices() {
        let items = this.selected_items.map(item => ({
            item_code: item.item_code,
            qty: item.qty
        }));
        
        frappe.call({
            method: 'pricing.pricing.page.quick_quotation.quick_quotation.get_multiple_items_prices',
            args: { items: items },
            callback: (r) => {
                if (r.message && r.message.items) {
                    this.selected_items = r.message.items;
                    this.render_items_table();
                    this.update_totals();
                }
            }
        });
    }
    
    update_totals() {
        let totals = {
            cost: 0,
            selling: 0,
            profit: 0,
            installment: 0,
            down_payment: 0,
            monthly_payment: 0,
            interest: 0,
            months: 0
        };
        
        this.selected_items.forEach(item => {
            let prices = item.prices || {};
            let installment = item.installment || {};
            
            totals.cost += prices.total_cost || 0;
            totals.selling += prices.total_selling_price || 0;
            totals.profit += prices.total_profit || 0;
            totals.installment += installment.total_amount || 0;
            totals.down_payment += installment.down_payment || 0;
            totals.monthly_payment += installment.monthly_payment || 0;
            totals.interest += installment.total_interest || 0;
            if (installment.number_of_months) totals.months = installment.number_of_months;
        });
        
        this.page.main.find('#total-cost').text(this.format_currency(totals.cost));
        this.page.main.find('#total-selling').text(this.format_currency(totals.selling));
        this.page.main.find('#total-profit').text(this.format_currency(totals.profit));
        this.page.main.find('#total-installment').text(this.format_currency(totals.installment));
        
        // نمایش جزئیات قسط
        if (totals.installment > 0) {
            this.page.main.find('#installment-details').show();
            this.page.main.find('#detail-down-payment').text(this.format_currency(totals.down_payment));
            this.page.main.find('#detail-monthly-payment').text(this.format_currency(totals.monthly_payment));
            this.page.main.find('#detail-months').text(totals.months);
            this.page.main.find('#detail-total-interest').text(this.format_currency(totals.interest));
        } else {
            this.page.main.find('#installment-details').hide();
        }
    }
    
    clear_all() {
        frappe.confirm('آیا می‌خواهید همه کالاها را پاک کنید؟', () => {
            this.selected_items = [];
            this.render_items_table();
            this.update_totals();
        });
    }
    
    create_quotation() {
        if (this.selected_items.length === 0) {
            frappe.show_alert({
                message: 'هیچ کالایی انتخاب نشده است',
                indicator: 'orange'
            });
            return;
        }
        
        let d = new frappe.ui.Dialog({
            title: 'ایجاد پیش‌فاکتور',
            fields: [
                {
                    fieldname: 'customer',
                    fieldtype: 'Link',
                    options: 'Customer',
                    label: 'مشتری'
                },
                {
                    fieldname: 'price_type',
                    fieldtype: 'Select',
                    label: 'نوع قیمت',
                    options: 'selling_price\nwith_markup\ninstallment',
                    default: 'selling_price'
                }
            ],
            primary_action_label: 'ایجاد',
            primary_action: (values) => {
                frappe.call({
                    method: 'pricing.pricing.page.quick_quotation.quick_quotation.create_quotation_from_items',
                    args: {
                        items: this.selected_items.map(i => ({
                            item_code: i.item_code,
                            qty: i.qty
                        })),
                        customer: values.customer,
                        price_type: values.price_type
                    },
                    callback: (r) => {
                        if (r.message && r.message.success) {
                            frappe.show_alert({
                                message: r.message.message,
                                indicator: 'green'
                            });
                            frappe.set_route('Form', 'Quotation', r.message.quotation);
                        }
                    }
                });
                d.hide();
            }
        });
        d.show();
    }
    
    print_quotation() {
        let me = this;
        let print_html = `
            <div style="direction: rtl; font-family: Tahoma, Arial; padding: 20px;">
                <h2 style="text-align: center; border-bottom: 2px solid #333; padding-bottom: 10px;">
                    پیشنهاد قیمت
                </h2>
                <p style="text-align: left;">تاریخ: ${frappe.datetime.nowdate()}</p>
                
                <table style="width: 100%; border-collapse: collapse; margin-top: 20px;">
                    <thead>
                        <tr style="background: #333; color: white;">
                            <th style="padding: 10px; border: 1px solid #333;">ردیف</th>
                            <th style="padding: 10px; border: 1px solid #333;">کد کالا</th>
                            <th style="padding: 10px; border: 1px solid #333;">نام کالا</th>
                            <th style="padding: 10px; border: 1px solid #333;">تعداد</th>
                            <th style="padding: 10px; border: 1px solid #333;">قیمت واحد</th>
                            <th style="padding: 10px; border: 1px solid #333;">قیمت کل</th>
                        </tr>
                    </thead>
                    <tbody>
        `;
        
        let total = 0;
        this.selected_items.forEach((item, index) => {
            let prices = item.prices || {};
            let unit_price = prices.unit_selling_price || 0;
            let total_price = prices.total_selling_price || 0;
            total += total_price;
            
            print_html += `
                <tr>
                    <td style="padding: 8px; border: 1px solid #ddd; text-align: center;">${index + 1}</td>
                    <td style="padding: 8px; border: 1px solid #ddd;">${item.item_code}</td>
                    <td style="padding: 8px; border: 1px solid #ddd;">${item.item_name || ''}</td>
                    <td style="padding: 8px; border: 1px solid #ddd; text-align: center;">${item.qty}</td>
                    <td style="padding: 8px; border: 1px solid #ddd; text-align: left;">${me.format_currency(unit_price)}</td>
                    <td style="padding: 8px; border: 1px solid #ddd; text-align: left;">${me.format_currency(total_price)}</td>
                </tr>
            `;
        });
        
        print_html += `
                    </tbody>
                    <tfoot>
                        <tr style="background: #f5f5f5; font-weight: bold;">
                            <td colspan="5" style="padding: 10px; border: 1px solid #ddd; text-align: left;">جمع کل:</td>
                            <td style="padding: 10px; border: 1px solid #ddd; text-align: left;">${me.format_currency(total)}</td>
                        </tr>
                    </tfoot>
                </table>
            </div>
        `;
        
        let w = window.open();
        w.document.write(print_html);
        w.document.close();
        w.print();
    }
    
    show_settings_dialog() {
        let d = new frappe.ui.Dialog({
            title: 'تنظیمات قیمت‌گذاری',
            fields: [
                {
                    fieldname: 'auto_price_list',
                    fieldtype: 'Link',
                    options: 'Auto Price List',
                    label: 'لیست قیمت خودکار'
                },
                {
                    fieldname: 'profit_margin',
                    fieldtype: 'Percent',
                    label: 'درصد سود',
                    default: 20
                },
                {
                    fieldname: 'target_discount_percentage',
                    fieldtype: 'Percent',
                    label: 'درصد تخفیف هدف'
                }
            ],
            primary_action_label: 'ذخیره',
            primary_action: (values) => {
                this.settings = values;
                d.hide();
                this.refresh_item_prices();
            }
        });
        d.show();
    }
    
    format_currency(value) {
        if (!value) return '0 ریال';
        return new Intl.NumberFormat('fa-IR').format(Math.round(value)) + ' ریال';
    }
}

File: pricing/pricing/page/quick_quotation/quick_quotation.py
# Copyright (c) 2024, Frappe Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, cint, today, getdate
import math

@frappe.whitelist()
def get_item_price_details(item_code, price_list=None, qty=1):
    """
    دریافت جزئیات قیمت یک کالا
    شامل قیمت نقدی، قسطی، با تخفیف و غیره
    """
    if not item_code:
        return {"error": "کد کالا وارد نشده است"}
    
    qty = flt(qty) or 1
    
    # دریافت اطلاعات کالا
    item = frappe.db.get_value("Item", item_code, 
        ["item_name", "item_group", "brand", "standard_rate", "valuation_rate"],
        as_dict=True
    )
    
    if not item:
        return {"error": "کالا یافت نشد"}
    
    # دریافت آخرین Auto Price List فعال
    latest_apl = get_latest_auto_price_list(price_list)
    
    # دریافت قیمت از Auto Price List Item
    apl_item = None
    if latest_apl:
        apl_item = frappe.db.get_value("Auto Price List Item",
            {"parent": latest_apl.name, "item_code": item_code},
            ["total_cost", "selling_price", "profit_amount", "final_price_with_markup",
             "down_payment_amount", "monthly_payment", "total_interest", "total_installment_amount",
             "commission_amount", "net_profit_after_commission"],
            as_dict=True
        )
    
    # محاسبه قیمت‌ها
    result = {
        "item_code": item_code,
        "item_name": item.item_name,
        "item_group": item.item_group,
        "brand": item.brand,
        "qty": qty,
        "prices": {}
    }
    
    if apl_item:
        # قیمت‌های محاسبه شده از Auto Price List
        unit_cost = flt(apl_item.total_cost)
        unit_selling_price = flt(apl_item.selling_price)
        unit_profit = flt(apl_item.profit_amount)
        
        result["prices"]["unit_cost"] = unit_cost
        result["prices"]["total_cost"] = unit_cost * qty
        result["prices"]["unit_selling_price"] = unit_selling_price
        result["prices"]["total_selling_price"] = unit_selling_price * qty
        result["prices"]["unit_profit"] = unit_profit
        result["prices"]["total_profit"] = unit_profit * qty
        result["prices"]["profit_percentage"] = (unit_profit / unit_cost * 100) if unit_cost > 0 else 0
        
        # قیمت با افزایش (برای امکان تخفیف)
        if apl_item.final_price_with_markup:
            result["prices"]["unit_price_with_markup"] = flt(apl_item.final_price_with_markup)
            result["prices"]["total_price_with_markup"] = flt(apl_item.final_price_with_markup) * qty
        
        # جزئیات قسط
        if apl_item.monthly_payment:
            result["installment"] = {
                "down_payment": flt(apl_item.down_payment_amount) * qty,
                "monthly_payment": flt(apl_item.monthly_payment) * qty,
                "total_interest": flt(apl_item.total_interest) * qty,
                "total_amount": flt(apl_item.total_installment_amount) * qty,
                "number_of_months": latest_apl.number_of_months if latest_apl else 0,
                "monthly_interest_rate": latest_apl.monthly_interest_rate if latest_apl else 0
            }
        
        # کمیسیون
        if apl_item.commission_amount:
            result["commission"] = {
                "amount": flt(apl_item.commission_amount) * qty,
                "net_profit": flt(apl_item.net_profit_after_commission) * qty
            }
        
        # تنظیمات از APL
        if latest_apl:
            result["settings"] = {
                "price_list": latest_apl.price_list,
                "profit_margin": latest_apl.profit_margin,
                "target_discount_percentage": latest_apl.target_discount_percentage,
                "required_markup_percentage": latest_apl.required_markup_percentage,
                "enable_installment": latest_apl.enable_installment,
                "number_of_months": latest_apl.number_of_months,
                "monthly_interest_rate": latest_apl.monthly_interest_rate,
                "down_payment_percentage": latest_apl.down_payment_percentage
            }
    else:
        # استفاده از قیمت استاندارد کالا
        unit_price = flt(item.standard_rate) or flt(item.valuation_rate) or 0
        result["prices"]["unit_selling_price"] = unit_price
        result["prices"]["total_selling_price"] = unit_price * qty
        result["warning"] = "این کالا در لیست قیمت خودکار نیست"
    
    return result

@frappe.whitelist()
def get_multiple_items_prices(items):
    """
    دریافت قیمت چندین کالا به صورت همزمان
    items: لیست از دیکشنری‌ها با item_code و qty
    """
    import json
    if isinstance(items, str):
        items = json.loads(items)
    
    results = []
    totals = {
        "total_cost": 0,
        "total_selling_price": 0,
        "total_profit": 0,
        "total_with_markup": 0,
        "total_installment": 0,
        "total_commission": 0
    }
    
    for item in items:
        item_code = item.get("item_code")
        qty = flt(item.get("qty", 1))
        
        if item_code:
            price_details = get_item_price_details(item_code, qty=qty)
            if "error" not in price_details:
                results.append(price_details)
                
                # جمع کل‌ها
                prices = price_details.get("prices", {})
                totals["total_cost"] += prices.get("total_cost", 0)
                totals["total_selling_price"] += prices.get("total_selling_price", 0)
                totals["total_profit"] += prices.get("total_profit", 0)
                totals["total_with_markup"] += prices.get("total_price_with_markup", 0)
                
                if price_details.get("installment"):
                    totals["total_installment"] += price_details["installment"].get("total_amount", 0)
                
                if price_details.get("commission"):
                    totals["total_commission"] += price_details["commission"].get("amount", 0)
    
    return {
        "items": results,
        "totals": totals
    }

@frappe.whitelist()
def search_items(search_term, limit=20):
    """
    جستجوی کالاها برای autocomplete
    """
    if not search_term:
        return []
    
    items = frappe.db.sql("""
        SELECT 
            item_code, 
            item_name, 
            item_group, 
            brand,
            standard_rate
        FROM `tabItem`
        WHERE disabled = 0
          AND (
              item_code LIKE %(search)s
              OR item_name LIKE %(search)s
              OR barcode LIKE %(search)s
          )
        LIMIT %(limit)s
    """, {
        "search": f"%{search_term}%",
        "limit": cint(limit)
    }, as_dict=True)
    
    return items

@frappe.whitelist()
def get_price_lists():
    """دریافت لیست‌های قیمت موجود"""
    return frappe.get_all("Price List", 
        filters={"enabled": 1},
        fields=["name", "price_list_name", "currency"],
        order_by="name"
    )

@frappe.whitelist()
def get_auto_price_lists():
    """دریافت لیست‌های قیمت خودکار"""
    return frappe.get_all("Auto Price List",
        filters={"docstatus": 1},
        fields=["name", "price_list", "valid_from", "valid_until", "profit_margin"],
        order_by="valid_from desc"
    )

def get_latest_auto_price_list(price_list=None):
    """دریافت آخرین Auto Price List فعال"""
    filters = {
        "docstatus": 1,
        "valid_from": ["<=", today()],
        "valid_until": [">=", today()]
    }
    
    if price_list:
        filters["price_list"] = price_list
    
    apl = frappe.db.get_value("Auto Price List", filters,
        ["name", "price_list", "profit_margin", "commission_percentage",
         "target_discount_percentage", "required_markup_percentage",
         "enable_installment", "number_of_months", "monthly_interest_rate",
         "down_payment_percentage", "enable_deferred_payment",
         "deferred_payment_months", "deferred_payment_interest_rate"],
        as_dict=True,
        order_by="valid_from desc"
    )
    
    return apl

@frappe.whitelist()
def calculate_custom_price(base_price, profit_margin=0, discount_percentage=0, 
                           enable_installment=False, number_of_months=0, 
                           monthly_interest_rate=0, down_payment_percentage=0, qty=1):
    """
    محاسبه قیمت سفارشی با پارامترهای دلخواه
    """
    base_price = flt(base_price)
    profit_margin = flt(profit_margin)
    discount_percentage = flt(discount_percentage)
    qty = flt(qty) or 1
    
    # محاسبه قیمت با سود
    selling_price = base_price * (1 + profit_margin / 100)
    profit_amount = selling_price - base_price
    
    # محاسبه قیمت با افزایش (برای تخفیف)
    price_with_markup = selling_price
    discount_amount = 0
    if discount_percentage > 0:
        markup_percentage = (discount_percentage / (100 - discount_percentage)) * 100
        price_with_markup = selling_price * (1 + markup_percentage / 100)
        discount_amount = price_with_markup * discount_percentage / 100
    
    result = {
        "unit_cost": base_price,
        "unit_selling_price": selling_price,
        "unit_profit": profit_amount,
        "unit_price_with_markup": price_with_markup,
        "unit_discount_amount": discount_amount,
        "profit_percentage": (profit_amount / base_price * 100) if base_price > 0 else 0,
        "markup_percentage": (discount_percentage / (100 - discount_percentage)) * 100 if discount_percentage < 100 else 0,
        "total_cost": base_price * qty,
        "total_selling_price": selling_price * qty,
        "total_price_with_markup": price_with_markup * qty
    }
    
    # محاسبه قسط
    if enable_installment and number_of_months > 0:
        monthly_rate = flt(monthly_interest_rate) / 100
        months = cint(number_of_months)
        down_pct = flt(down_payment_percentage) / 100
        
        total_price = price_with_markup * qty
        down_payment = total_price * down_pct
        financed_amount = total_price - down_payment
        
        if monthly_rate > 0:
            # فرمول بهره مرکب
            monthly_payment = financed_amount * (monthly_rate * (1 + monthly_rate)**months) / ((1 + monthly_rate)**months - 1)
            total_installment = down_payment + (monthly_payment * months)
            total_interest = total_installment - total_price
        else:
            monthly_payment = financed_amount / months
            total_installment = total_price
            total_interest = 0
        
        result["installment"] = {
            "down_payment": down_payment,
            "monthly_payment": monthly_payment,
            "number_of_months": months,
            "total_interest": total_interest,
            "total_amount": total_installment,
            "interest_percentage": ((1 + monthly_rate) ** months - 1) * 100
        }
    
    return result

@frappe.whitelist()
def create_quotation_from_items(items, customer=None, price_type="selling_price"):
    """
    ایجاد پیش‌فاکتور از آیتم‌های انتخاب شده
    """
    import json
    if isinstance(items, str):
        items = json.loads(items)
    
    # ایجاد Quotation
    quotation = frappe.new_doc("Quotation")
    quotation.quotation_to = "Customer" if customer else "Lead"
    if customer:
        quotation.party_name = customer
    
    for item in items:
        item_code = item.get("item_code")
        qty = flt(item.get("qty", 1))
        
        # دریافت قیمت
        price_details = get_item_price_details(item_code, qty=1)
        if "error" not in price_details:
            prices = price_details.get("prices", {})
            
            # انتخاب نوع قیمت
            if price_type == "selling_price":
                rate = prices.get("unit_selling_price", 0)
            elif price_type == "with_markup":
                rate = prices.get("unit_price_with_markup", 0)
            elif price_type == "installment":
                installment = price_details.get("installment", {})
                rate = installment.get("total_amount", 0) / qty if qty > 0 else 0
            else:
                rate = prices.get("unit_selling_price", 0)
            
            quotation.append("items", {
                "item_code": item_code,
                "qty": qty,
                "rate": rate
            })
    
    quotation.insert()
    
    return {
        "success": True,
        "quotation": quotation.name,
        "message": f"پیش‌فاکتور {quotation.name} ایجاد شد"
    }

